// hmc-shake/shake.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ID = exports.Sleep = exports.on = exports.Random = exports.nextTick = exports.ShakeList = exports.GetSize = exports.has = exports.del = exports.remove = exports.Const = exports.set = exports.get = exports.isset = exports.wait = exports.shake = void 0;
/**所有防抖数据都会在这里存放 */
const ConstShakeList = new Set;
const handleShakeEventList = [];
/**
 * 删除键值
 * @param Key 键值
 * @param Timeout 延迟执行时间 没有则直接删除
 * @returns
 */
function RemoveShakeKey(Key, Timeout, CallBack) {
    if (Timeout) {
        setTimeout(() => {
            ConstShakeList.delete(Key);
            if (CallBack)
                CallBack();
            for (const handleShake of handleShakeEventList)
                if (handleShake.key == Key)
                    handleShake.CallBack.apply(ConstShakeList, ["remove", Timeout]);
        }, Timeout);
        return true;
    }
    ConstShakeList.delete(Key);
    for (const handleShake of handleShakeEventList)
        if (handleShake.key == Key)
            handleShake.CallBack.apply(ConstShakeList, ["remove", Timeout]);
    return true;
}
/**
 * 添加防抖key
 * @param Key 键值
 * @param Timeout 超时时间 没有则永久
 */
function SetShakeKey(Key, Timeout, CallBack) {
    ConstShakeList.add(Key);
    if (Timeout)
        RemoveShakeKey(Key, Timeout, CallBack);
    for (const handleShake of handleShakeEventList)
        if (handleShake.key == Key)
            handleShake.CallBack.apply(ConstShakeList, ["add", Timeout]);
}
/**
 * 判断是否有这个Key
 */
const HasShakeKey = (key) => {
    for (const handleShake of handleShakeEventList)
        if (handleShake.key == key)
            handleShake.CallBack.apply(ConstShakeList, ["has", undefined]);
    return ConstShakeList.has(key);
};
/**
 * 已有的防抖则返回 true  没有则添加
 * @param Key 键值
 * @param Timeout 超时时间
 */
function HasIfNotAdd(Key, Timeout, CallBack) {
    if (HasShakeKey(Key))
        return true;
    SetShakeKey(Key, Timeout, CallBack);
    return false;
}
/**
* 随机数值
* @param {*} Min 最小
* @param {*} Max 最大
*/
function GetRandomNum(Min, Max) {
    var Range = Max - Min;
    var Rand = Math.random();
    return (Min + Math.round(Rand * Range));
}
/**
 * 已有的防抖则返回 true  没有则添加 并随机延时
 * @param Key 键值
 * @param MinTimeout 最小超时时间
 * @param MaxTimeout 最大超时时间
 */
function Random(Key, MinTimeout, MaxTimeout) {
    if (Array.isArray(MinTimeout)) {
        MaxTimeout = MinTimeout[1];
        MinTimeout = MinTimeout[0];
    }
    let Timeout = GetRandomNum(MinTimeout || 500, MaxTimeout || 5000);
    if (HasShakeKey(Key))
        return true;
    SetShakeKey(Key, Timeout);
    return false;
}
exports.Random = Random;
/**
 * 监听事件冒泡
 * @param Key
 * @param type
 * @param CallBack
 */
function on(Key, type, CallBack) {
    if (Key && type && !!CallBack) {
        handleShakeEventList.push({
            key: Key,
            type: type,
            CallBack: CallBack
        });
    }
}
exports.on = on;
/**
 * 没有该键值的时候添加 到时间后返回异步成功 存在时返回异步失败
 * @param key 键值
 * @param ms 延迟
 * @returns
 */
function nextTick(key, ms) {
    return new Promise((resolve, reject) => {
        if (HasShakeKey(key))
            return reject();
        HasIfNotAdd(key, ms, () => resolve());
    });
}
exports.nextTick = nextTick;
/**
 * 所有键的数量
 * @returns
 */
const GetShakeKeySize = () => { return ConstShakeList.size; };
/**
 * 已有的防抖则返回 true  没有则添加
 * @param Key 键值
 * @param Timeout 超时时间
 */
let isset = HasIfNotAdd;
exports.isset = isset;
/**判断是否有这个Key */
let get = HasShakeKey;
exports.get = get;
/**
 * 添加防抖key
 * @param Key 键值
 * @param Timeout 超时时间 没有则永久
 */
let set = SetShakeKey;
exports.set = set;
/**
 * 固定该键值
 * @param Key 键值
 * @returns
 */
let Const = (Key) => SetShakeKey(Key);
exports.Const = Const;
/**
 * 删除一个键值
 * @param Key 键值
 * @param Timeout 延迟执行时间 没有则直接删除
 * @param CallBack 删除成功后发生的回调
 * @returns
 */
let remove = RemoveShakeKey;
exports.remove = remove;
/**
 * 删除一个键值
 * @param Key 键值
 * @param Timeout 延迟执行时间 没有则直接删除
 * @param CallBack 删除成功后发生的回调
 * @returns
 */
let del = RemoveShakeKey;
exports.del = del;
/**判断是否有这个Key */
let has = HasShakeKey;
exports.has = has;
/**所有键值数量 */
let GetSize = GetShakeKeySize;
exports.GetSize = GetSize;
/** 所有防抖数据都会在这里存放*/
let ShakeList = ConstShakeList;
exports.ShakeList = ShakeList;
/**
 * 异步阻塞
 * @param ms
 * @returns
 */
function Sleep(ms) {
    let FormatMs = isNaN(Number(ms)) ? 150 : Number(ms);
    return new Promise((resolve) => setTimeout(resolve, Number(FormatMs)));
}
exports.Sleep = Sleep;
function wait(key, IgnoreTime, awitTitme, CallBack) {
    let _CallBack = typeof awitTitme == "function" ? awitTitme : typeof CallBack == "function" ? CallBack : null;
    if (Array.isArray(key))
        key = key.join('::');
    if (ConstShakeList.has(key)) {
        return false;
    }
    ConstShakeList.add(key);
    let _Promise = new Promise(async function (resolve, reject) {
        await Sleep(IgnoreTime);
        if (typeof awitTitme == "number")
            await Sleep(awitTitme);
        ConstShakeList.delete(typeof key == "string" ? key : key.join("::"));
        _CallBack && _CallBack();
        resolve(undefined);
    });
    return _CallBack ? true : _Promise;
}
exports.wait = wait;
/**
 * 随机的文本id
 * @returns
 */
function ID() { return (Date.now().toString(36) + Math.random().toString(36).slice(2, 7)).toUpperCase(); }
exports.ID = ID;
exports.shake = {
    wait, isset, get, set, Const, remove, del, has, GetSize, ShakeList, nextTick, Random, on, Sleep, ID
};
exports.default = exports.shake;
